### PR TITLE
don't call asyncio.run for async handling

### LINKED ISSUE #93: asyncio functionality is mostly unusable after async handling was applied
Reporting on my own failures here. In #90 we added automatic `async` handling and did so by using `asyncio.run`:

https://github.com/smarie/mkdocs-gallery/blob/8a16972a3d1dd9728a5406bb83f4fe631b9ad4c6/src/mkdocs_gallery/gen_single.py#L782-L783

The problem is, and I didn't know this before, that `asyncio.run` is meant to be only called once. This is what the [documentation](https://docs.python.org/3/library/asyncio-runner.html#asyncio.run) says about this:

> This function should be used as a main entry point for asyncio programs, and should ideally only be called once.

However, we call it for every code block that needs `async` handling. That in itself is not an issue as the gallery here builds just fine. Unfortunately, it breaks down when other libraries that are part of the documentation building also need to use some `async` functionality.

For example, [`panel`](https://panel.holoviz.org/) creates an `asyncio.Lock` on import. When building some `async` galleries before the import, we effectively have the following situation

```python
import asyncio
import asyncio.events

async def foo():
    pass

asyncio.run(foo())

loop = asyncio.events.get_event_loop()
```

And the `get_event_loop` fails with

```
RuntimeError: There is no current event loop in thread 'MainThread'.
```

because we have previously used `asyncio.run`.

My absolute crude workaround for now is the following snippet that has to be placed before any further `asyncio` functionality is used:

```python
import asyncio

asyncio.get_event_loop_policy()._local._set_called = False
```

Please note that this works for my case, but I have no idea if this screws up other cases as we are messing with low-level private attributes here.

I'll look on how to restructure the async handling to avoid using `asyncio.run` to avoid this mess all together.

### PR DESCRIPTION
**Pull request by pmeier**
This Pull request fixes #93

 - [x] I completed the issue numbers (#xx) in the sentence above. The word "fixes" should remain in front of each issue
 - [x] My PR is tagged as draft when I'm still working on it, and I remove the draft flag when it is ready for review.
- [ ] I added one or several `changelog.md` entries in the appropriate "in progress" section (not the last release one)

### b - My PR fixes some issues:

*(Delete this section if not relevant)*

 - [x] Each issue is well-described, well-labeled, and contains reproducible code examples.

@smarie This PR needs cleanup, but I wanted to get early feedback from you if the overall functionality is what you want.

I think it would make sense to move all the async stuff to a dedicated module or put them into `utils`. Any opinion here?

### CODE REVIEW

**Comment by smarie** [comment]
Thanks @pmeier for this nice contribution.
The issue description thread looks good and the proposed solution as well.

I think that it would be a good idea, if not too difficult, to have a specific test in your test file reproducing the issue #93 explicitly, so that it can ensure non-regression in the future.

Indeed some `utils_asyncio.py` module could be a good idea, if enough material is to go there. But don't rush into this, you can also keep using utils.py if the amount of code for async is not too much. I let you decide what is most convenient for you.

**Comment by smarie** [comment]
@pmeier do you by any chance have any progress on this ? Let me know

**Comment by pmeier** [comment]
TBH, I completely forgot about this. Unfortunately, I'm swamped for the next month and will only have some dedicated time for this in October. 

> I think that it would be a good idea, if not too difficult, to have a specific test in your test file reproducing the issue #93 explicitly, so that it can ensure non-regression in the future.

That is already implemented.

**Comment by pmeier** [comment]
@smarie I just did it now instead punting it to later. PTAL.

File: examples/plot_12_async.py [hunk 1]
```diff
@@ -31,13 +31,7 @@ async def afn():
...
 # %%
 # Without any handling, the snippet above would trigger a `SyntaxError`, since we are using `await` outside of an
-# asynchronous context. With the handling, it works just fine.
-#
-# The background handling will only be applied if it is actually needed. Meaning, you can still run your asynchronous
-# code manually if required.
-
-asyncio.run(afn())
-
+# asynchronous context. With the background handling, it works just fine.
```
Comment (smarie) [inline examples/plot_12_async.py:35]: Is the empty cell between those comments sections needed ? otherwise please merge the two

**Comment by smarie** [comment]
Amazing ! I just made a minor comment. Could you please also add a changelog entry in a new 0.10.3 section ? https://github.com/smarie/mkdocs-gallery/blob/main/docs/changelog.md

I'll release right after
Thanks !

**Comment by smarie** [comment]
Thanks @pmeier  !

**Pull request merged**